### PR TITLE
Update `azurerm_linux|windows_virtual_machine` and `azurerm_linux|windows_virtual_machine_scale_set` - Support `encryption_at_host_enabled`

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Supersede #8134

### CODE REVIEW

**Comment by ArcturusZhang** [comment]
Tests are passing:
```
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled (318.19s)
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate (349.75s)
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK (601.90s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost (241.76s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate (324.41s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK (589.60s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
--- PASS: TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled (283.23s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate (417.15s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled (631.12s)
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled (388.23s)
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate (415.94s)
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK (547.36s)

**Comment by ArcturusZhang** [comment]
All tests have passed:
```
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabled (287.18s)
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledUpdate (418.35s)
=== RUN   TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK
--- PASS: TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK (567.91s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHost (235.12s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostUpdate (305.41s)
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherEncryptionAtHostWithCMK (590.23s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled
--- PASS: TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabled (357.87s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledUpdate (782.34s)
=== RUN   TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK
--- PASS: TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK (794.18s)
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabled (346.02s)
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledUpdate (465.22s)
=== RUN   TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
=== PAUSE TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
=== CONT  TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK
--- PASS: TestAccAzureRMWindowsVirtualMachineScaleSet_otherEncryptionAtHostEnabledWithCMK (700.09s)
PASS
```

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

File: azurerm/internal/services/compute/windows_virtual_machine_scale_set_resource.go [hunk 1]
```diff
@@ -235,9 +240,9 @@ func resourceArmWindowsVirtualMachineScaleSet() *schema.Resource {
 				ForceNew: true,
 				Default:  string(compute.Manual),
 				ValidateFunc: validation.StringInSlice([]string{
-					string(compute.Automatic),
+					string(compute.AutomaticByOS),
 					string(compute.Manual),
-					string(compute.Rolling),
+					string(compute.AutomaticByPlatform),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/windows_virtual_machine_scale_set_resource.go:245]: these should be:

```suggestion
 					string(compute.UpgradeModeAutomatic),
 					string(compute.UpgradeModeManual),
 					string(compute.UpgradeModeRolling),
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
The tests have picked up a bug here - if we can fix that up then this should otherwise be good to merge 👍

**Comment by ArcturusZhang** [comment]
Hi @tombuildsstuff I have resolved the last comment, could you please have a look again?

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Comment by WodansSon** [comment]
image

**Pull request merged**